diff --git a/bin/core.d.ts b/bin/core.d.ts
index 9d7f6fb8bf7fd305f5ffa51f929198d1d3e2aa46..bd6d1f0f8e7c87b86695a8ab7e581a00d816d167 100644
--- a/bin/core.d.ts
+++ b/bin/core.d.ts
@@ -76,6 +76,7 @@ export interface Drawable {
     sets: OpSet[];
 }
 export interface PathInfo {
+    type: OpSetType;
     d: string;
     stroke: string;
     strokeWidth: number;
diff --git a/bin/generator.js b/bin/generator.js
index e28cc61c25fc597b30fd0bc57da028235b6e2055..a46b6f4adb3ad5970d0fb3592b35ea2ce91b441c 100644
--- a/bin/generator.js
+++ b/bin/generator.js
@@ -26,6 +26,7 @@ export class RoughGenerator {
             disableMultiStroke: false,
             disableMultiStrokeFill: false,
             preserveVertices: false,
+            fixedDecimalPlaceDigits: 3
         };
         this.config = config || {};
         if (this.config.options) {
@@ -208,6 +209,7 @@ export class RoughGenerator {
             switch (drawing.type) {
                 case 'path':
                     path = {
+                        type: 'path',
                         d: this.opsToPath(drawing),
                         stroke: o.stroke,
                         strokeWidth: o.strokeWidth,
@@ -216,6 +218,7 @@ export class RoughGenerator {
                     break;
                 case 'fillPath':
                     path = {
+                        type: 'fillPath',
                         d: this.opsToPath(drawing),
                         stroke: NOS,
                         strokeWidth: 0,
@@ -238,6 +241,7 @@ export class RoughGenerator {
             fweight = o.strokeWidth / 2;
         }
         return {
+            type: 'fillSketch',
             d: this.opsToPath(drawing),
             stroke: o.fill || NOS,
             strokeWidth: fweight,